### ISSUE TITLE
Remove the /cluster directory

### ISSUE THREAD

**Issue by timothysc** [kubernetes/kubernetes#78995]
For years we have publicly stated that the /cluster directory is deprecated and not maintained.  However every cycle it is updated and there are bugs found+fixed by sig-cluster-lifecycle.  

I'd like to enumerate in this issue what needs to get done in order for us to wholesale remove the /cluster directory. 

/assign @dims @spiffxp @justinsb @timothysc  
/cc @liggitt @neolit123

**Comment by neolit123** [comment]
or potentially break it down and move certain sub-folders out of tree.
currently it contains a collection of items:
- image files (conformance image, etcd)
- kube-up
- random bash scripts used by test-infra/kubetest deployers as helpers
- addons and the addon manager.

**Comment by liggitt** [comment]
looking at the references to it...

/sig testing
for e2e bringup

/sig scalability
for kubemark bringup

**Comment by sftim** [comment]
Is this also relevant to SIG Docs?

**Comment by timothysc** [comment]
xref -https://github.com/kubernetes/kubernetes/pull/78543 
^ Is an example of continued technical debt that we see and have to pay for in different ways across SCL.

**Comment by timothysc** [comment]
xref -https://github.com/kubernetes/kubernetes/pull/78543 
^ Is an example of continued technical debt that we see and have to pay for in different ways across SCL.

**Comment by jaypipes** [comment]
> Is this also relevant to SIG Docs?

Yes, I think this is a good example of where `/cluster` scripts are referenced from the docs: https://github.com/kubernetes/website/pull/14929

**Comment by andrewsykim** [comment]
For v1.16: investigate if cluster API is a potential replacement and meets the same level of coverage as `/cluster` and enumerate what is missing cluster API

/assign @alejandrox1

**Comment by dims** [comment]
Step 1 we talked about was ... look at all the CI jobs that use the kube-up in the cluster directory and inventory the knobs/settings/configurations they setup or use and cross check if cluster-api/kubeadm allows us to do the same.
Step 2 mirror the GCE e2e CI Job (`pull-kubernetes-e2e-gce` to be specific) using cluster-api for AWS.

Both steps can be done in parallel and will need help/effort/coordination on the wg-k8s-infra and sig-testing folks

**Comment by alejandrox1** [comment]
I'll start with step 1 right away.
So step 2 is free if anyone wants to work on this as well 😃

**Comment by dims** [comment]
@alejandrox1 sounds good. please start a google doc or something that we can use to compile notes on the various flags/options

**Comment by alejandrox1** [comment]
On the google doc: https://docs.google.com/document/d/1p3c_sOALbEzg2VH2OPz3w9yKwlwaU4jATyS0lqxFzt4/edit?usp=sharing

Still got a lot to do for step 1 but will ping when this is ready 
/cc @mariantalla

**Comment by mariantalla** [comment]
Could I start work on step 2 (i.e. investigating/starting the refactor of `pull-kubernetes-e2e-gce` to use Cluster API)?

I'll assign myself, but please feel free to unassign me if someone else is already working on it!

/assign

**Comment by justaugustus** [comment]
Go for it, @mariantalla! :)

**Comment by dims** [comment]
yes please @mariantalla !

**Comment by neolit123** [comment]
i have create a ticket for tracking replacing the kube-up jobs with Cluster API:
https://github.com/kubernetes/kubernetes/issues/82532

cc @mariantalla

**Comment by dims** [comment]
Update: CAPG now as a presubmit and a periodic job that starts up a cluster and runs the e2e/conformance tests

https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api-provider-gcp#e2e%20tests
https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api-provider-gcp#pr-e2e

**Comment by pjh** [comment]
> I'd like to enumerate what needs to get done in order for us to wholesale remove the /cluster directory.

Clusters with Windows nodes on GCE depend on the `cluster/gce/` kube-up code. This dependency was noted in https://github.com/kubernetes/kubernetes/issues/82532#issue-491700946, but I don't see Windows mentioned in this issue yet or in the doc @alejandrox1 mentioned  above.

That other issue says:

> this is one of the trickier items actually, since CAPI does not support Windows.

**Comment by neolit123** [comment]
>  since CAPI does not support Windows.

**Comment by markyjackson-taulia** [comment]
@spiffxp @timothysc @justinsb @dims @mariantalla Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on November 18. Is this issue still intended for 1.17?

**Comment by jtslear** [comment]
Hello,
Bug Triage team here for the 1.18 release. It appears that linked PR https://github.com/kubernetes/kubernetes/issues/82532 has been bumped to milestone 1.19.  I'm proceeding to do the same on this issue.  Should anyone disagree with this, please feel free to re-assign.

/milestone v1.19

**Comment by spiffxp** [comment]
/milestone clear
I don't think we're committing to this for v1.19

**Comment by neolit123** [comment]
PRs to move the conformance image:
https://github.com/kubernetes/kubernetes/pull/93937
https://github.com/kubernetes/kubernetes/pull/93937

cc @dims @BenTheElder

**Comment by BenTheElder** [comment]
commented already 🙃
<https://emojipedia.org/upside-down-face/#:~:text=A%20classic%20smiley%2C%20turned%20upside,sense%20of%20goofiness%20or%20silliness.&text=Upside%2DDown%20Face%20was%20approved,to%20Emoji%201.0%20in%202015.>

**Comment by cheftako** [comment]
/assign
/triage accepted
cc @jpbetz

**Comment by jpbetz** [comment]
The etcd images are not kubernetes specific. The main thing they do is automatically upgrade etcd one minor at a time, per etcd administration guidelines, when the cluster administrator upgrades etcd to a new version. E.g. if the cluster administrator upgrades to etcd 3.4 and the cluster is currently on 3.1, it upgrades first to 3.2 and then to 3.3 before upgrading to 3.4.

So I'm tempted to ask if the etcd community would be willing to own this. If the etcd community was okay with this, the main issue to solve is that the images are published to the k8s.io container repo.

cc @gyuho, @ptabor, @wenjiaswe

**Comment by gyuho** [comment]
> The etcd images are not kubernetes specific

Yeah, I think we can implement some mechanism to update docs and container images in the registry for Kubernetes as part of etcd release process.

@ptabor @wenjiaswe Any thoughts?

**Comment by justinsb** [comment]
We could also do this in the etcdadm project, as that is a kubernetes-sigs project and is thus set up to push to k8s repos / follow k8s governance etc.

**Comment by ptabor** [comment]
The northstar IMHO is that we should get rid of process of 'updating' etcd by running consecutive minor versions of etcd 
and instead etcd should have a dedicated tool: e.g. `etcdstoragectl migrate` (I would call it etcdadm, but the name is taken ;) ) process  that  knows the DB changes between different versions and explicitly 'fixes' the database instead of depending on the running for undefined duration full etcd server. 

There are multiple benefits of this: 
   - Such code is 'hands-on' documentation to the storage format differences. 
   - the tool/library should be incorporated into backup-restore process
   - no maintenance of fragile multi-version generation code. 
   - if changes are related to WAL log, the tool should create a new snapshot and truncate history. Etcd doesn't to this on boot. 

Maybe its naive, but in scope of etcd-v3 minor versions, during work on [etcd storage-format documentation](https://docs.google.com/document/d/1O2o1IApHWmSioXG3fez4eVlUHOrXICYGNVIzaqNS0IQ/edit?resourcekey=0-e6Iywgdkol0uiVBAaV1oww) I haven't spotted any significant differences in the format. 
It's more like: 'set a default' if given field is missing, style of O(5) rules.

**Comment by jpbetz** [comment]
> The northstar IMHO is that we should get rid of process of 'updating' etcd by running consecutive minor versions of etcd

I'm a big fan of this approach. If we can get support for this one the etcd side then the solution w/r/t the `/cluster` direct might just to be drop these images and use upstream etcd.

Curious what others think.

**Comment by gyuho** [comment]
> w/r/t the /cluster direct might just to be drop these images and use upstream etcd

Yeah always wonder why we even need two separate etcd registries. I am open to dropping the whole container support in our release process, and let downstream project build their own. Or merge onto Kubernetes-community managed registry.

**Comment by aojea** [comment]
I think that what has to be done is to have an alternative as reliable as cluster or better, I'm not saying that cluster is great, but I didn't see anything better so far ... and this is very easy to measure with testgrid

**Comment by dims** [comment]
Sketching up requirements:
https://hackmd.io/pw1kt61lRM-wZh5MU1G_QQ

Here's a snapshot as of Feb27 just to be safe, please see all the comments on the hackmd.
```
# Requirements for cluster/kube-up.sh replacement

We have a lot of choices in the community like Kops/Kubespray/CAPA etc but all of them take too long or do not support all test sceanarios or hard to inject freshly built code. Hence the search for a new replacement. As you can see from this is a [long standing issue](https://github.com/kubernetes/kubernetes/issues/78995) and one that open to "easy" fixes.

- Must support 80% of jobs today (revisit all the environment flags we use to control different aspects of the cluster to verify)
- All nodes must run on a VM to replicate (how it is being done today, we already have `kind` to replicate things run inside a container)
- Must be able to deploy the cluster built directly from either a PR or the tip of a branch (to cover both presubmit and periodic jobs)
- Must use `kubeadm` to bootstrap the both the control plane node and the worker nodes
- `kubeadm` needs systemd for running `kubelet` so the images deployed should use systemd
- Must have a mechanical way to translate existing jobs to this new harness
- Should have a minimum of moving parts to ensure we are not chasing flakes and digging into things we don't need to
- Should have a clean path (UX) to debug things like we have today (logs from VM/cloudinit/systemd/kubelet/containers should tell the whole story)
- Should work on GCP and AWS, with Azure closely behind (so should be pluggable for other clouds)
- Should not be a hack, we need this to sustain us for 5-8 years at least. 
- Should work with things we already have like prow+boskos
- Adding a new service that is always on (like prow/boskos) should be well thought out as it is not trivial to debug yet another thing. Any such solution must be rock solid.
- we should be able to test kubeadm, external cloud providers (CPI) and storage drivers (CSI). 
- should support switching between containerd and CRI-O as well as cgroup v1/v2

We would need this solution to run in parallel for a full release cycle and offer equal or better results. This will need active owners who can take care of it for the long haul as worst case we will drop it like a hot potato however elegant or technologically superior the solution is.

Forward looking ideas that can be explored:
- Making whatever changes are needed to those tools to boot faster / allow injection of code.
```

**Comment by BenTheElder** [comment]
> - Must use `kubeadm` to bootstrap the both the control plane node and the worker nodes

I'm not sure this should be a hard requirement, as much appreciation as I have for kubeadm, it does a relatively small portion of cluster bootstrapping and we often need to test things it does not do / reconfigure things it does not directly support configuring anyhow. I'd suggest using kubeadm if  we staffed writing something new from scratch, but I don't think we should preclude pre-existing options based on it.

> We have a lot of choices in the community like Kops/Kubespray/CAPA etc but all of them take too long or do not support all test sceanarios or hard to inject freshly built code. Hence the search for a new replacement. As you can see from this is a [long standing issue](https://github.com/kubernetes/kubernetes/issues/78995) and one that open to "easy" fixes.


I'm also not sure the statement about kOps in particular is accurate, we actually used to run kOps on Kubernetes PRs as recently as 2019 and it worked fine. We lost this due to the AWS bill going unpaid somewhere between Amazon and the CNCF leading to the account being terminated and no ability to run the job.  https://github.com/kubernetes/kubernetes/issues/73444#issuecomment-458397963

We didn't spin up kops-GCE instead because we already had cluster-up and there was a strong push for Cluster-API, but kops-GCE works and passes tests. It's also relatively mature.
https://testgrid.k8s.io/kops-gce#kops-gce-latest

kubespray does not support kubernetes @ HEAD and CAPA is not passing tests / adds overhead like the local cluster, AIUI.

kops is already supported by kubetest + kubetest2, boskos, etc. and is passing tests on GCE + AWS. In theory it supports other providers but I don't think we have any CI there yet.

> - Must have a mechanical way to translate existing jobs to this new harness

This seems impractical. Most of the tricky part here is jobs setting kube-up environment variables that reconfigure the cluster. A human will have to dive into the scripts and see what each env ultimately does and then remap it to any other tool used. Unless that tool is also written in bash, we won't have drop in identical behavior, a lot of it is expanding variables inside bash-generated configuration files.

We could target the most important jobs and work from there.

**Comment by aojea** [comment]
@upodroid brought a related topic today, he is working on adding ARM64 support to the CI and modifyng the cluster scripts to do that https://github.com/kubernetes/kubernetes/pull/120144

I can see kops already is running arm64 (@justinsb) https://testgrid.k8s.io/google-aws#kops-aws-arm64-ci
I rather prefer these news CI jobs to use a new tool instead of keep building on the `cluster` folder, if possible, or we'll never break this loop

**Comment by upodroid** [comment]
I'll create the arm64 equivalent of https://testgrid.k8s.io/sig-release-master-blocking#gce-ubuntu-master-containerd using kops and then have a conversation about migrating the amd64 one to kops.

**Comment by neolit123** [comment]
> I'll create the arm64 equivalent of https://testgrid.k8s.io/sig-release-master-blocking#gce-ubuntu-master-containerd on GCE using kops and then have a conversation about migrating the amd64 one to kops.

for k0ps related questions we have #kops-dev on k8s slack.

**Comment by dims** [comment]
@upodroid how close are we to doing this? 1.31 perhaps?

**Comment by dims** [comment]
@BenTheElder no objections!

**Comment by BenTheElder** [comment]
Ok I'll fire off something later, I think we should roll up a project to wind down:
- `kubetest` (the legacy one in test-infra not https://sigs.k8s.io/kubetest2 and kubetest2-kops etc)
- `scenarios/kubernetes_e2e.py` (test-infra wrapper script)
- `cluster/` (excluding local-up-cluster.sh which we'll sustain longer term?)
Along with reviving https://github.com/kubernetes/enhancements/issues/2464 or https://github.com/kubernetes/enhancements/issues/4224

Even if we don't get everything completed and can't commit to an alpha/beta/GA timeline, I think we need a written plan somewhere, it's a little too adhoc at the moment. https://github.com/kubernetes/enhancements/issues/4224 probably came closest previously.

There's some other details we should discuss like get-kube.sh, I think a "process KEP" outlining the intent is probably the best approximation for those details.

**Comment by upodroid** [comment]
One of the main blockers in winding down cluster/ is the presence of cloud-provider-gcp tests in k/k. We have some release-blocking jobs that test some of that functionality.(That should be tested in k/cloud-provider-gcp and everything in k/k should be doable on any cluster provisioner)

https://github.com/kubernetes/kubernetes/tree/master/test/e2e/cloud/gcp

Can SIG Cloud Provider +1 their removal? It would accelerate completing this project

**Comment by elmiko** [comment]
wow, i really kicked the anthill...

if this directory is still in use that's totally fine by me, we were just triaging old issues and this came up.

@jpbetz yes, i do believe there was a time when there were scripts here used by the node lifecycle test (for example), but i think all of those linkages have been removed. plus we are exploring a different direction for future tests that would keep the provider specific bits closer to the ccm implementations.

perhaps we should close this issue after figuring out the ownership stuff (or open a separate issue for that)?

@timothysc 's original request is from many years ago and the repo looked much different at that time, plus we were just beginning the journey of removing all cloud provider specific stuff.

**Comment by elmiko** [comment]
> One of the main blockers in winding down cluster/ is the presence of cloud-provider-gcp tests in k/k. We have some release-blocking jobs that test some of that functionality.(That should be tested in k/cloud-provider-gcp and everything in k/k should be doable on any cluster provisioner)
> 
> https://github.com/kubernetes/kubernetes/tree/master/test/e2e/cloud/gcp
> 
> Can SIG Cloud Provider +1 their removal? It would accelerate completing this project

i would like a little more time to study this, but from my recent audit i think some of that functionality is getting skipped by using a fake provider in the test framework. i'm happy to spend some time reviewing and proposing its removal if that is where the investigation leads to.

**Comment by elmiko** [comment]
> Ok I'll fire off something later, I think we should roll up a project to wind down:

i'm happy to volunteer time to removing the cloud provider tests.

**Comment by soltysh** [comment]
> * `cluster/` (excluding local-up-cluster.sh which we'll sustain longer term?)
>       Along with reviving [kubetest2 CI migration enhancements#2464](https://github.com/kubernetes/enhancements/issues/2464) or [Replace kube-up e2e clusters with kops for kubernetes/kubernetes e2e testing enhancements#4224](https://github.com/kubernetes/enhancements/issues/4224)

If the majority of the purpose of this directory is to support our tests, why not move the necessary bits under `hack/` directory? We already have there `hack/e2e-internal/` which unsurprisingly invoke scripts from `cluster/` directory. Why not just combine them all under a single place and thus allowing us to remove the  `cluster/` directory entirely, and better expressing the scripts intent?

**Comment by liggitt** [comment]
I think relocating it would ripple into many more places than you expect, I would rather see us spend that time and energy burning down or converting uses to other frameworks and repos

**Comment by dims** [comment]
> I think relocating it would ripple into many more places than you expect, I would rather see us spend that time and energy burning down or converting uses to other frameworks and repos

💯 agree with this approach.

**Comment by elmiko** [comment]
> I think relocating it would ripple into many more places than you expect, I would rather see us spend that time and energy burning down or converting uses to other frameworks and repos

from the perspective of the cloud provider tests this is something i have been exploring with a colleague, we are working towards a demonstration for the SIG. i tend to agree that it will be better time spent developing the new path, at least for the tests under `test/e2e/cloud`.

**Comment by jasminekeller751-commits** [comment]
agree — happy to volunteer some time to help remove the cloud provider
tests.
Should we start a mini-project or just coordinate under the SIG to track
progress?

**Comment by elmiko** [comment]
> Should we start a mini-project or just coordinate under the SIG to track progress?

for the cloud provider specific tests, i would prefer to coordinate through the SIG.

**Comment by BenTheElder** [comment]
> I think relocating it would ripple into many more places than you expect, I would rather see us spend that time and energy burning down or converting uses to other frameworks and repos

Yeah. And it ripples into some things we'd also like to just phase out versus modify further (like some of the scripts and tools in other repos mentioned above ... kubetest etc)
